pkg/handlers/generic/mutation/parallelimagepulls/inject.go

```diff
@@ -0,0 +1,174 @@
+// Copyright 2025 Nutanix. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package parallelimagepulls
+
+import (
+	"bytes"
+	"context"
+	_ "embed"
+	"fmt"
+	"text/template"
+
+	apiextensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	bootstrapv1 "sigs.k8s.io/cluster-api/bootstrap/kubeadm/api/v1beta1"
+	controlplanev1 "sigs.k8s.io/cluster-api/controlplane/kubeadm/api/v1beta1"
+	runtimehooksv1 "sigs.k8s.io/cluster-api/exp/runtime/hooks/api/v1alpha1"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/nutanix-cloud-native/cluster-api-runtime-extensions-nutanix/api/v1alpha1"
+	"github.com/nutanix-cloud-native/cluster-api-runtime-extensions-nutanix/common/pkg/capi/clustertopology/handlers/mutation"
+	"github.com/nutanix-cloud-native/cluster-api-runtime-extensions-nutanix/common/pkg/capi/clustertopology/patches"
+	"github.com/nutanix-cloud-native/cluster-api-runtime-extensions-nutanix/common/pkg/capi/clustertopology/patches/selectors"
+	"github.com/nutanix-cloud-native/cluster-api-runtime-extensions-nutanix/common/pkg/capi/clustertopology/variables"
+)
+
+const (
+	// VariableName is the external patch variable name.
+	VariableName = "maxParallelImagePullsPerNode"
+
+	kubeletConfigurationPatchFilePath = "/etc/kubernetes/patches/kubeletconfigurationmaxparallelimagepulls+strategic.json"
+)
+
+var (
+	//go:embed embedded/kubeletconfigpatch.yaml
+	kubeletConfigPatchYAML []byte
+
+	kubeletConfigPatchTemplate = template.Must(template.New("kubeletConfigPatch").Parse(string(kubeletConfigPatchYAML)))
+)
+
+type maxParallelImagePullsPerNode struct {
+	variableName      string
+	variableFieldPath []string
+}
+
+func NewPatch() *maxParallelImagePullsPerNode {
+	return newMaxParallelImagePullsPerNodePatch(
+		v1alpha1.ClusterConfigVariableName,
+		VariableName,
+	)
+}
+
+func newMaxParallelImagePullsPerNodePatch(
+	variableName string,
+	variableFieldPath ...string,
+) *maxParallelImagePullsPerNode {
+	return &maxParallelImagePullsPerNode{
+		variableName:      variableName,
+		variableFieldPath: variableFieldPath,
+	}
+}
+
+func (h *maxParallelImagePullsPerNode) Mutate(
+	ctx context.Context,
+	obj *unstructured.Unstructured,
+	vars map[string]apiextensionsv1.JSON,
+	holderRef runtimehooksv1.HolderReference,
+	_ client.ObjectKey,
+	clusterGetter mutation.ClusterGetter,
+) error {
+	log := ctrl.LoggerFrom(ctx).WithValues(
+		"holderRef", holderRef,
+	)
+
+	maxParallelImagePullsPerNode, err := variables.Get[int32](
+		vars,
+		h.variableName,
+		h.variableFieldPath...,
+	)
+	switch {
+	case err != nil && !variables.IsNotFoundError(err):
+		return err
+	case variables.IsNotFoundError(err):
+		log.V(5).Info("max parallel image pulls is not set, using default value of 1")
+		maxParallelImagePullsPerNode = 1
```

Comment:
This will cause an unwanted Cluster rollout because this file will now always be written out. Should we remove the default in the API and leave the defaulting to Kubernetes?
Otherwise we need to update the handler versions and only include this in the new one.

Reply:
Updated the logic to prevent unwanted cluster rollout on CAREN upgrade.